### PR TITLE
org.virt_manager.virt-viewer

### PR DESCRIPTION
**Pull request by AdrianKoshka**
This is a rough-draft PR, I still need to write an appstream file, but mostly everything else seems done.

### CODE REVIEW

File: org.virt-manger.virt-viewer.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<!-- https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html -->
```
[the remark this answers is not in the document]
Comment (AdrianKoshka) [reply org.virt-manger.virt-viewer.appdata.xml]: Not yet, I'm waiting for upstream to reply to an email I sent to their mailing-list. I'll probably submit this appstream rough-draft to upstream in a following email.

File: org.virt-manger.virt-viewer.json [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+                    "sha256": "f893f32b1ef6d09dd23cda39b8a2567be7c2aebda58026288e1362f042e20808"
+                }
+            ]
+        },
+        {
+            "name": "spice-protocol",
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "git://anongit.freedesktop.org/spice/spice-protocol",
```
[the remark this answers is not in the document]
Comment (AdrianKoshka) [reply org.virt-manger.virt-viewer.json]: Will-do

**Comment by AdrianKoshka** [comment]
I had been building only `remote-viewer` before because I didn't know I also needed to build `libvirt` to build `virt-viewer`, I'm working on fixing this now. what I intend to happen is when you run `flatpak run org.virt_manager.virt-viewer`, `virt-viewer` is ran, but if you run the `.desktop` of `remote-viewer`, then `remote-viewer` is ran.

**Comment by AdrianKoshka** [comment]
I put the wrong https link for that git repo, will push a fix for that too, as soon as I get libvirt to build.

**Comment by AdrianKoshka** [comment]
Oh, thanks, I'll fix this.

**Comment by AdrianKoshka** [comment]
This still needs some serious cleanup, but it should build correctly now.

File: org.virt_manger.virt-viewer.appdata.xml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  <summary>A remote desktop client built with GTK-VNC, SPICE-GTK and libvirt</summary>
+  <!-- https://specifications.freedesktop.org/menu-spec/latest/apa.html -->
+  <!-- https://specifications.freedesktop.org/menu-spec/latest/apas02.html -->
+  <categories>
+    <category>Utility</category>
+    <categories>Network</categories>
+  </categories>
+  <url type="homepage">https://virt-manager.org/download/</url>
+  <url type="bugtracker">https://bugzilla.redhat.com/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&classification=Community&component=virt-viewer&product=Virtualization%20Tools</url>
+  <url type="help">https://support.mozilla.org/questions/thunderbird</url>
```
Comment (eszlari) [inline org.virt_manger.virt-viewer.appdata.xml]: Wrong help URL.

File: org.virt_manger.virt-viewer.appdata.xml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+<!-- https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html -->
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
```
Comment (eszlari) [inline org.virt_manger.virt-viewer.appdata.xml]: No screenshots.

File: org.virt_manger.virt-viewer.appdata.xml (hunk 3 above)
Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.appdata.xml]: Thanks for catching this, I wasn't really able to find a good "help" URL, so I'll just remove that line. Is that okay?

File: org.virt_manger.virt-viewer.json [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+                    "tag": "v2.5.3",
+                    "commit": "aaa5251791bf0b1640afcba77a7d79ea23c42d53"
+                }
+            ]
+        },
+        {
+            "name": "lz4",
+            "build-options": {
+                "env": {
+                    "PREFIX": "/app"
```
Comment (eszlari) [inline org.virt_manger.virt-viewer.json]: You use ${FLATPAK_DEST} above.

File: org.virt_manger.virt-viewer.json [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+                }
+            },
+            "buildsystem": "simple",
+            "build-commands": [
+                "make",
```
Comment (eszlari) [inline org.virt_manger.virt-viewer.json]: `make -j $FLATPAK_BUILDER_N_JOBS`

File: org.virt_manger.virt-viewer.json [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+            ]
+        },
+        {
+            "name": "gcrypt-wrapper",
+            "buildsystem": "simple",
+            "cleanup": [
+                "*"
+            ],
+            "build-commands": [
+                "install -Dm755 libgcrypt-config ${FLATPAK_DEST}/bin/libgcrypt-config"
```
Comment (eszlari) [inline org.virt_manger.virt-viewer.json]: Can be shortened to: `install libgcrypt-config ${FLATPAK_DEST}/bin`

File: org.virt_manger.virt-viewer.appdata.xml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  <!-- From the GNOME Software utility -->
+  <summary>A remote desktop client built with GTK-VNC, SPICE-GTK and libvirt</summary>
+  <!-- https://specifications.freedesktop.org/menu-spec/latest/apa.html -->
+  <!-- https://specifications.freedesktop.org/menu-spec/latest/apas02.html -->
+  <categories>
+    <category>Utility</category>
+    <categories>Network</categories>
+  </categories>
+  <url type="homepage">https://virt-manager.org/download/</url>
+  <url type="bugtracker">https://bugzilla.redhat.com/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&classification=Community&component=virt-viewer&product=Virtualization%20Tools</url>
```
Comment (eszlari) [inline org.virt_manger.virt-viewer.appdata.xml]:
```
$ appstream-util validate-relax org.virt_manger.virt-viewer.appdata.xml 
org.virt_manger.virt-viewer.appdata.xml: failed to parse org.virt_manger.virt-viewer.appdata.xml: Error on line 16: Entity did not end with a semicolon; most likely you used an ampersand character without intending to start an entity - escape ampersand as &amp;
```

File: org.virt_manger.virt-viewer.appdata.xml (hunk 3 above)
Comment (eszlari) [reply org.virt_manger.virt-viewer.appdata.xml]: Sure.

File: org.virt_manger.virt-viewer.json (hunk 7 above)
Comment (eszlari) [reply org.virt_manger.virt-viewer.json]: Sorry, that was wrong. You could only drop the "m755".

**Comment by nedrichards** [comment]
bot, build bot, build org.virt_manager.virt-viewer

**Comment by AdrianKoshka** [comment]
So now I've gotten `virt-viewer` in a working state, the next problem I need to tackle is getting `virt-viewer` to connect to `qemu:///system`. `qemu:///system` is the URI for qemu/libvirt running on that local machine.

I think this would need to be done by passing a file through to the sandbox via `--filesystem=` (because `virt-viewer` looks for a file which is a socket). I'll also start working on bettering the appstream file.

File: org.virt_manger.virt-viewer.json (hunk 5 above)
Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.json]: So are you saying to replace `PREFIX` with `FLATPAK_DEST`?

Comment (eszlari) [reply org.virt_manger.virt-viewer.json]: No, see module `gcrypt-wrapper`. I meant: Use `/app` or `${FLATPAK_DEST}` consistently.

Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.json]: does https://github.com/flathub/flathub/pull/391/commits/72a693627838a6fd053fc4dbb4274fe8f8afe268 solve this?

Comment (eszlari) [reply org.virt_manger.virt-viewer.json]: Sure. Not that it does have any effect, it was only a minor style thing.

Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.json]: Thanks, you've been doing a great job of helping me better my manifest.

**Comment by nedrichards** [comment]
bot, build org.virt_manager.virt-viewer

**Comment by AdrianKoshka** [comment]
Sorry for the radio silence, I was on vacation and went to an event over the past weekend.

**Comment by nedrichards** [comment]
No worries at all! "Having A Life" is not a problem, when you're donating your time for free you don't owe anybody anything (unless you want to).

**Comment by AdrianKoshka** [comment]
I don't think I'll be able to make USB redirection work, mostly because almost all of `/dev` is read-only.

> - All mounts are read-only, except:
>   - root tmpfs (e.g. for tmp subdir and transient home)
>   - /var
>   - /proc (except /proc/sys, proc/sysrq-trigger, /proc/irq and /proc/bus which are 
>  read-only)
>   - /dev/pts
>   - /dev/shm


Another thing I haven't worked out is how to connect to `qemu:///system`, which is qemu(kvm)/libvirtd running on the host. `virt-viewer` seemingly needs `/var/run/libvirt/libvirt-sock-ro`.

File: org.virt_manger.virt-viewer.appdata.xml (hunk 4 above)
Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.appdata.xml]: What resolution should screenshots be? As a reference.

**Comment by nedrichards** [comment]
bot, build org.virt_manager.virt-viewer

File: org.virt_manger.virt-viewer.json [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+                "/share/gtk-doc",
+                "/share/man",
+                "/share/pkgconfig",
+                "*.la",
+                "*.a"
+            ],
+            "sources": [
+                {
+                    "type": "archive",
+                    "url": "https://github.com/gentoo/eudev/archive/v3.2.5.tar.gz",
```
Comment (barthalion) [inline org.virt_manger.virt-viewer.json]: We have it in shared-modules: https://github.com/flathub/shared-modules/tree/master/udev

File: org.virt_manger.virt-viewer.json [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+                "--without-storage-rbd",
+                "--without-storage-sheepdog",
+                "--without-storage-gluster",
+                "--without-storage-zfs",
+                "--without-storage-vstorage"
+            ],
+            "sources": [
+                {
+                    "type": "archive",
+                    "url": "https://libvirt.org/sources/libvirt-4.3.0.tar.xz",
```
Comment (barthalion) [inline org.virt_manger.virt-viewer.json]: Not sure if it matters for app functionality, but 4.4.0 is o ut.

File: org.virt_manger.virt-viewer.json [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+    "finish-args" : [
+        "--share=ipc",
+        "--socket=x11",
+        "--socket=wayland",
+        "--socket=pulseaudio",
+        "--share=network",
+        "--talk-name=org.a11y.*",
+        "--device=all",
+        "--filesystem=xdg-pictures",
+        "--filesystem=~/.ssh",
```
Comment (barthalion) [inline org.virt_manger.virt-viewer.json]: There is also `--socket=ssh-auth` now, but I'm not sure if it's enough for ssh to work.

File: org.virt_manger.virt-viewer.json (hunk 9 above)
Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.json]: I think the version in shared modules was out of date for my needs.

File: org.virt_manger.virt-viewer.json (hunk 11 above)
Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.json]: I plan on testing this tonight.

Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.json]: So, while the ssh-auth socket is enough for ssh to work on a basic level, if `~/.ssh` isn't passed through, it seemingly doesn't create a `known_hosts` file, and will always prompt the user whether they trust the fingerprint or not.

File: org.virt_manger.virt-viewer.json (hunk 10 above)
Comment (AdrianKoshka) [reply org.virt_manger.virt-viewer.json]: Updated it (and a few other things) anyway.

**Comment by nedrichards** [comment]
bot, build org.virt_manager.virt-viewer

**Comment by AdrianKoshka** [comment]
Thanks.

**Comment by barthalion** [comment]
bot, build org.virt_manager.virt-viewer

File: Makefile [hunk 12]
```diff
@@ -0,0 +1,18 @@
+TARGET_REPO = repo
+FLATPAK_BUILDER = $(shell which flatpak-builder)
+MANIFEST = org.virt_manger.virt-viewer.json
```
Comment (barthalion) [inline Makefile]: Here is also a typo. I'd rather remove `Makefile` completely.

File: org.virt_manager.virt-viewer.appdata.xml [hunk 13]
```diff
@@ -0,0 +1,74 @@
+<!-- https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html -->
```
Comment (barthalion) [inline org.virt_manager.virt-viewer.appdata.xml]: You are not installing this file anywhere. Have you submitted it upstream too?

Comment (AdrianKoshka) [reply org.virt_manager.virt-viewer.appdata.xml]: Haven't yet, the last time I tried to contact upstream I got absolutely nothing back.

**Comment by nedrichards** [comment]
bot, build org.virt_manager.virt-viewer

**Comment by nedrichards** [comment]
bot, build org.virt_manager.virt-viewer

**Comment by AdrianKoshka** [comment]
Not happy with how this flatpak operates. Mostly do USB redirection not working. I don't think I'll be submitting it after-all. I guess if anyone wants to pick it up in future, I could help them though. Thanks for all the help, apologies for the wasted effort.

**Pull request closed without merging**